### PR TITLE
Issue #3792: Workaround for php floating point precision problem

### LINKED ISSUE #3792: Simple decimals fail to pass validation
**Description of the bug**
I am running into a problem where a number/decimal field is not allowing values that should pass the validation.

It is possibly related to [this Drupal 8 issue](https://www.drupal.org/project/drupal/issues/2230909)? Not certain but sounds similar, and the steps to reproduce listed here also reproduce the error on a Backdrop install: 

> **Problem/Motivation**
> 
> - Create decimal field, set precision to 10 (minimum in the UI and scale to 4
> - Saving new node with value 19999.0000 succeeds (precision is 5+4 = 9).
> - Saving new node with value 99999.0000 fails (same precision as above).
> 
> or
> 
> - Create decimal field, set scale to anything over 6 (need 8 to store bitcoin values for example). I used 16 as precission.
> - Saving new node with value 20.12345678 fails validation while 0.1234567 succeeds.
> 
> This is because Drupal\Component\Utility\Number::validStep() returns false. Detailed investigation on this function (which is only used once in Drupal), reveals that the first argument ($value) is received as a string, but $step and $offset are floats. PHP seems to slightly mangle the $step value on the first case above from 0.0001 to 0.00010000000000000001438. Passing the step parameter as a string in the case of decimal numbers maintains the correct precision, and allows a correct approximation calculation.
>
> **Proposed resolution**
> 
> Bypass weak PHP floating-point handling by passing the step as a string.
> 

**Update**

After some experiments with a much simpler, then the D8 approach, it turned out that neither of these works properly.

It needs a small change in function number_field_widget_form() and a complete rewrite of function valid_number_step().

Also note that several of the existing tests were wrong and needed switching, see [this comment](https://github.com/backdrop/backdrop-issues/issues/3792#issuecomment-724819508) for details and explanation.

Several additional automated tests have been added to verify the new approach.

---
There are currently two PR, which might be confusing.

- by @indigoxela https://github.com/backdrop/backdrop/pull/3392
- by @BWPanda outdated, just fixes a few false negatives

### PR DESCRIPTION
**Pull request by indigoxela**
Fixes https://github.com/backdrop/backdrop-issues/issues/3792

### CODE REVIEW

File: core/includes/common.inc [hunk 1]
```diff
@@ -1450,32 +1450,59 @@ function valid_url($url, $absolute = FALSE) {
...
+
+  // Both numbers are integers - use modulo right away.
+  if ((int) $value == $value && (int) $step == $step && $value < PHP_INT_MAX) {
+    return $value % $step == 0;
+  }
+
+  // Convert from possible exponent notation to decimal strings to get the scale
+  // values. Note: sprintf also rounds values but seems to give more reliable
+  // results than number_format.
+  // "10" is the maximum value allowed by the number widget, so higher scale
```
Comment (klonos) [inline core/includes/common.inc:1475]: > "10" is the maximum value allowed by the number widget,

I wonder why this is the case here 🤔

File: core/includes/common.inc [hunk 2]
```diff
@@ -1450,32 +1450,59 @@ function valid_url($url, $absolute = FALSE) {
...
+  $max_scale = 13;
+  $step_string = rtrim(sprintf("%.{$max_scale}F", $step), '0');
+  $value_string = rtrim(sprintf("%.{$max_scale}F", $value), '0');
+
+  // If the actual value has more significant decimals than expected, it has a
+  // higher precision than desired and is not divisible by the step.
+  // But if the scale value high, and the length of string (minus comma) is
+  // beyond the php setting for precision, it's more likely that the conversion
+  // to string caused a problem with rounding. It's a compromise to allow it,
+  // because there's no reliably way for php to handle it.
```
Comment (klonos) [inline core/includes/common.inc]: Should either be: `no reliable way` or `no way for php to handle it reliably.`

File: core/includes/common.inc (hunk 1 above)
Comment (indigoxela) [reply core/includes/common.inc:1475]: It's in the [settings form](https://github.com/backdrop/backdrop/blob/c6a99af4723ec52cfb35151af762c3efab85f7b8/core/modules/field/modules/number/number.module#L59)

I couldn't tell why :wink: , but 10 is the max for this settings form item.

File: core/includes/common.inc (hunk 2 above)
Comment (indigoxela) [reply core/includes/common.inc]: :+1: and I also saw that an "is" is missing above.

**Comment by quicksketch** [comment]
Thanks! Merged into 1.x and 1.17.x.

**Pull request merged**